File: src/ui/widgets/composite/simple_options_flyout.py
import logging
import sys

from PyQt6.QtCore import QEasingCurve, QPoint, QPropertyAnimation, Qt, pyqtSignal
from PyQt6.QtGui import (
    QBrush,
    QColor,
    QFont,
    QFontMetrics,
    QGuiApplication,
    QPainter,
    QPen,
)
from PyQt6.QtWidgets import (
    QApplication,
    QGraphicsDropShadowEffect,
    QHBoxLayout,
    QLabel,
    QVBoxLayout,
    QWidget,
)

from core.constants import AppConstants
from src.shared_toolkit.ui.managers.theme_manager import ThemeManager

logger = logging.getLogger("ImproveImgSLI")

class _SimpleRow(QWidget):
    clicked = pyqtSignal(int)

    def __init__(self, index: int, text: str, is_current: bool, item_height: int, item_font: QFont, parent: QWidget = None):
        super().__init__(parent)
        self.index = index
        self.text = text
        self.is_current = is_current
        self._hovered = False
        self.theme_manager = ThemeManager.get_instance()
        self.setFixedHeight(item_height)
        self.setMouseTracking(True)
        layout = QHBoxLayout(self)
        layout.setContentsMargins(10, 0, 10, 0)
        self.label = QLabel(text)
        self.label.setFont(item_font)
        layout.addWidget(self.label)
        try:
            self.theme_manager.theme_changed.connect(self._apply_label_style)
        except Exception:
            pass
        self._apply_label_style()

    def _apply_label_style(self):
        tm = self.theme_manager
        text_color_key = "list_item.text.normal"
        font = QFont(self.label.font())
        font.setBold(False)
        self.label.setFont(font)
        self.label.setStyleSheet(f"color: {tm.get_color(text_color_key).name()}; background: transparent;")

    def enterEvent(self, e):
        self._hovered = True
        self.update()
        super().enterEvent(e)

    def leaveEvent(self, e):
        self._hovered = False
        self.update()
        super().leaveEvent(e)

    def mouseReleaseEvent(self, e):
        if e.button() == Qt.MouseButton.LeftButton and self.rect().contains(e.pos()):
            self.clicked.emit(self.index)
        super().mouseReleaseEvent(e)

    def paintEvent(self, e):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)
        tm = self.theme_manager
        if self.is_current or self._hovered:
            bg_color = tm.get_color("list_item.background.hover")
        else:
            bg_color = tm.get_color("list_item.background.normal")
        painter.setPen(Qt.PenStyle.NoPen)
        painter.setBrush(QBrush(bg_color))
        background_rect = self.rect().adjusted(2, 2, -2, -2)
        painter.drawRoundedRect(background_rect, 5, 5)
        if self.is_current:
            indicator_pen = QPen(tm.get_color("accent"))
            indicator_pen.setWidth(3)
            indicator_pen.setCapStyle(Qt.PenCapStyle.RoundCap)
            painter.setPen(indicator_pen)
            y1, y2 = self.rect().top() + 7, self.rect().bottom() - 7
            x = self.rect().left() + indicator_pen.width()
            painter.drawLine(x, y1, x, y2)

class SimpleOptionsFlyout(QWidget):
    item_chosen = pyqtSignal(int)
    closed = pyqtSignal()

    SHADOW_RADIUS = 10
    MARGIN = 8
    APPEAR_EXTRA_Y = 6

    def __init__(self, app_ref):
        super().__init__(None)
        self.app_ref = app_ref
        self._theme = ThemeManager.get_instance()
        self._options: list[str] = []
        self._current_index: int = -1
        self._item_height = 36
        self._item_font = QFont(QApplication.font(self))
        self._move_duration_ms = AppConstants.FLYOUT_ANIMATION_DURATION_MS
        self._move_easing = QEasingCurve.Type.OutQuad
        self._drop_offset_px = 80
        self._anim: QPropertyAnimation | None = None

        if sys.platform in ('linux', 'darwin'):
            window_flags = Qt.WindowType.Popup | Qt.WindowType.FramelessWindowHint
        else:
            window_flags = Qt.WindowType.ToolTip | Qt.WindowType.FramelessWindowHint

        self.setWindowFlags(window_flags)
        self.setAttribute(Qt.WidgetAttribute.WA_TranslucentBackground)

        self.container_widget = QWidget(self)
        self.container_widget.setObjectName("FlyoutWidget")
        shadow = QGraphicsDropShadowEffect(self)
        shadow.setBlurRadius(self.SHADOW_RADIUS)
        shadow.setOffset(1, 2)
        shadow.setColor(QColor(0, 0, 0, 120))
        self.container_widget.setGraphicsEffect(shadow)

        self.main_layout = QVBoxLayout(self)
        self.main_layout.setContentsMargins(self.MARGIN, self.MARGIN, self.MARGIN, self.MARGIN)
        self.main_layout.addWidget(self.container_widget)

        self.container_layout = QVBoxLayout(self.container_widget)
        self.container_layout.setContentsMargins(4, 4, 4, 4)
        self.container_layout.setSpacing(2)

        self.content_widget = QWidget(self.container_widget)
        self.content_layout = QVBoxLayout(self.content_widget)
        self.content_layout.setContentsMargins(0, 0, 0, 0)
        self.content_layout.setSpacing(2)
        self.container_layout.addWidget(self.content_widget)

        self._apply_style()
        self._theme.theme_changed.connect(self._apply_style)
        self.hide()

    def _apply_style(self):
        tm = self._theme
        bg_color = tm.get_color("flyout.background").name(QColor.NameFormat.HexArgb)
        border_color = tm.get_color("flyout.border").name(QColor.NameFormat.HexArgb)
        self.container_widget.setStyleSheet(f"""
            QWidget#FlyoutWidget {{
                background-color: {bg_color};
                border: 1px solid {border_color};
                border-radius: 8px;
            }}
        """)

    def set_row_height(self, h: int): self._item_height = max(28, int(h))
    def set_row_font(self, f: QFont): self._item_font = QFont(f)

    def populate(self, labels: list[str], current_index: int = -1):
        self._options = list(labels)
        self._current_index = current_index if 0 <= current_index < len(self._options) else -1
        while self.content_layout.count():
            item = self.content_layout.takeAt(0)
            if w := item.widget():
                w.deleteLater()
            del item
        for i, text in enumerate(self._options):
            row = _SimpleRow(i, text, i == self._current_index, self._item_height, self._item_font, self.content_widget)
            row.clicked.connect(self._on_row_clicked)
            self.content_layout.addWidget(row)
        self._update_size()

    def _update_size(self):
        num = len(self._options)
        h_content = 50 if num == 0 else (num * self._item_height + max(0, num - 1) * self.content_layout.spacing())
        container_h = h_content + 8
        self.container_widget.setFixedHeight(container_h)
        width = max(self.container_widget.sizeHint().width(), 200)
        self.container_widget.setFixedWidth(width)
        self.setFixedSize(width + self.MARGIN * 2, container_h + self.MARGIN * 2)

    def show_below(self, anchor_widget: QWidget):
        if self._anim: self._anim.stop()

        anchor_width = anchor_widget.width()
        fm = QFontMetrics(self._item_font)
        max_text_width = max((fm.boundingRect(label).width() for label in self._options), default=0)

        content_width = max_text_width + 20 + 8 + 10
        desired_width = max(anchor_width, content_width, 180)

        try:
            anchor_pos = anchor_widget.mapToGlobal(anchor_widget.rect().bottomLeft())
            anchor_center_x = anchor_widget.mapToGlobal(anchor_widget.rect().center()).x()
            if hasattr(self, 'windowHandle') and anchor_widget.windowHandle():
                self.windowHandle().setTransientParent(anchor_widget.windowHandle())
        except Exception: return

        self.container_widget.setFixedWidth(desired_width)
        self.setFixedWidth(desired_width + self.MARGIN * 2)
        self.adjustSize()

        total_width, total_height = self.width(), self.height()
        final_y = anchor_pos.y() + self.APPEAR_EXTRA_Y - self.MARGIN
        final_x = int(anchor_center_x - total_width / 2)

        try:
            screen = anchor_widget.screen() or QGuiApplication.screenAt(anchor_pos)
            avail = screen.availableGeometry()
        except Exception: avail = QGuiApplication.primaryScreen().availableGeometry()

        final_x = max(avail.left(), min(final_x, avail.right() - total_width))
        final_y = max(avail.top(), min(final_y, avail.bottom() - total_height))

        start_pos, end_pos = QPoint(final_x, final_y - self._drop_offset_px), QPoint(final_x, final_y)
        self.move(start_pos)
        self.show()

        anim_pos = QPropertyAnimation(self, b"pos", self)
        anim_pos.setDuration(self._move_duration_ms)
        anim_pos.setStartValue(start_pos)
        anim_pos.setEndValue(end_pos)
        anim_pos.setEasingCurve(self._move_easing)

        anim_pos.finished.connect(self._on_animation_finished)
        self._anim = anim_pos
        anim_pos.start()

    def _on_animation_finished(self):
        if self._anim:

            anim_obj = self._anim
            self._anim = None
            anim_obj.deleteLater()

    def _on_row_clicked(self, idx: int):
        self.item_chosen.emit(idx)
        self.hide()

    def hideEvent(self, e):
        super().hideEvent(e)

        if self._anim:
            self._anim.stop()

        try:
            self.closed.emit()
        except Exception:
            pass

File: cpp/toolkit/src/composite/simple_options_flyout.cpp
#include "sli/toolkit/composite/simple_options_flyout.h"

#include <chrono>
#include <optional>

#include <QApplication>
#include <QEasingCurve>
#include <QEnterEvent>
#include <QFontMetrics>
#include <QFrame>
#include <QGuiApplication>
#include <QHBoxLayout>
#include <QHideEvent>
#include <QKeyEvent>
#include <QLabel>
#include <QMouseEvent>
#include <QPainter>
#include <QPaintEvent>
#include <QPen>
#include <QPropertyAnimation>
#include <QScrollArea>
#include <QScreen>
#include <QVBoxLayout>
#include <QWidget>

#include "sli/toolkit/atomic/minimalist_scrollbar.h"
#include "sli/toolkit/composite/flyout.h"
#include "sli/toolkit/theme.h"

namespace sli::toolkit {

// ─── SimpleRow ───────────────────────────────────────────────────────────────
// Internal row widget mirroring Python's _SimpleRow.
// Custom paintEvent handles _RowBackgroundLayer and _CurrentIndicatorLayer.

namespace {

class SimpleRow final : public QWidget {
  Q_OBJECT

 public:
  explicit SimpleRow(int index,
                     const QString& text,
                     bool isCurrent,
                     int itemHeight,
                     const QFont& itemFont,
                     QWidget* parent = nullptr)
      : QWidget(parent), index_(index), text_(text), isCurrent_(isCurrent) {
    setFixedHeight(itemHeight);
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
    setCursor(Qt::PointingHandCursor);
    setMouseTracking(true);

    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(10, 0, 10, 0);

    label_ = new QLabel(text, this);
    label_->setFont(itemFont);
    label_->setAttribute(Qt::WA_TransparentForMouseEvents, true);
    applyLabelStyle();
    layout->addWidget(label_);

    Theme::onThemeChanged(this, [this]() {
      applyLabelStyle();
      update();
    });
  }

  int rowIndex() const { return index_; }
  bool isCurrent() const { return isCurrent_; }
  void setIsCurrent(bool c) {
    if (isCurrent_ != c) {
      isCurrent_ = c;
      update();
    }
  }

 signals:
  void rowClicked(int index);

 protected:
  void paintEvent(QPaintEvent*) override {
    // _RowBackgroundLayer: inset rounded rect, list_item.background tokens
    const bool active = isCurrent_ || isHovered_;
    const QString key = active ? QStringLiteral("list_item.background.hover")
                               : QStringLiteral("list_item.background.normal");
    const QColor bgColor = Theme::getColor(key);

    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    // Inset by 2 on each side (.adjusted(2, 2, -2, -2))
    const QRect bgRect = rect().adjusted(2, 2, -2, -2);
    p.setPen(Qt::NoPen);
    p.setBrush(bgColor);
    p.drawRoundedRect(bgRect, 5, 5);

    // _CurrentIndicatorLayer: left accent line, only when is_current
    if (isCurrent_) {
      const QColor accentColor = Theme::getColor(QStringLiteral("accent"));
      QPen pen(accentColor);
      pen.setWidth(3);
      pen.setCapStyle(Qt::RoundCap);
      p.setPen(pen);
      const int x = rect().left() + pen.width();
      p.drawLine(x, rect().top() + 7, x, rect().bottom() - 7);
    }
  }

  void enterEvent(QEnterEvent* event) override {
    isHovered_ = true;
    update();
    QWidget::enterEvent(event);
  }

  void leaveEvent(QEvent* event) override {
    isHovered_ = false;
    update();
    QWidget::leaveEvent(event);
  }

  void mousePressEvent(QMouseEvent* event) override {
    if (event->button() == Qt::LeftButton) {
      isPressed_ = true;
      update();
    }
    QWidget::mousePressEvent(event);
  }

  void mouseReleaseEvent(QMouseEvent* event) override {
    if (event->button() == Qt::LeftButton && isPressed_) {
      isPressed_ = false;
      update();
      if (rect().contains(event->pos())) {
        emit rowClicked(index_);
      }
    }
    QWidget::mouseReleaseEvent(event);
  }

 private:
  void applyLabelStyle() {
    // Mirrors Python's _apply_label_style: font not bold, class "option-label"
    QFont f = label_->font();
    f.setBold(false);
    label_->setFont(f);
    label_->setProperty("class", QStringLiteral("option-label"));
    // Use dialog.text token for label color (standard text color in flyouts)
    const QColor textColor = Theme::getColor(QStringLiteral("dialog.text"));
    label_->setStyleSheet(
        QStringLiteral("color: %1; background: transparent;")
            .arg(textColor.name(QColor::HexArgb)));
  }

  int index_ = 0;
  QString text_;
  bool isCurrent_ = false;
  bool isHovered_ = false;
  bool isPressed_ = false;
  QLabel* label_ = nullptr;
};

}  // namespace

// We need the MOC to process SimpleRow defined inside an anonymous namespace
// inside a .cpp. Use #include at the bottom.

// ─── SimpleOptionsFlyout ──────────────────────────────────────────────────────

SimpleOptionsFlyout::SimpleOptionsFlyout(QWidget* parentWidget)
    : Flyout(parentWidget), parentWidget_(parentWidget) {

  setObjectName(QStringLiteral("sliSimpleOptionsFlyout"));

  // Python: self._item_font = QFont(QApplication.font(self))
  itemFont_ = QApplication::font(this);

  // Python: self.content_layout.setSpacing(0)
  //         self.content_layout.setContentsMargins(4, 4, 4, 4)
  contentLayout_->setSpacing(0);
  contentLayout_->setContentsMargins(4, 4, 4, 4);

  // Scroll area — mirrors Python setup
  scrollArea_ = new QScrollArea(this);
  scrollArea_->setWidgetResizable(true);
  scrollArea_->setFrameShape(QFrame::NoFrame);
  scrollArea_->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
  scrollArea_->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
  scrollArea_->setVerticalScrollBar(new MinimalistScrollBar(Qt::Vertical));
  scrollArea_->setStyleSheet(
      QStringLiteral("QScrollArea { background: transparent; border: none; }"));
  scrollArea_->viewport()->setAutoFillBackground(false);
  scrollArea_->viewport()->setStyleSheet(
      QStringLiteral("background: transparent;"));

  // Rows container
  rowsContainer_ = new QWidget();
  rowsContainer_->setStyleSheet(QStringLiteral("background: transparent;"));
  rowsLayout_ = new QVBoxLayout(rowsContainer_);
  rowsLayout_->setContentsMargins(0, 0, 0, 0);
  rowsLayout_->setSpacing(2);
  rowsLayout_->addStretch();

  scrollArea_->setWidget(rowsContainer_);
  contentLayout_->addWidget(scrollArea_);

  hide();
}

void SimpleOptionsFlyout::setMaxVisibleItems(int n) {
  maxVisibleItems_ = std::max(1, n);
}

void SimpleOptionsFlyout::setRowHeight(int h) {
  itemHeight_ = std::max(28, h);
}

void SimpleOptionsFlyout::setRowFont(const QFont& f) {
  itemFont_ = f;
}

void SimpleOptionsFlyout::populate(const QStringList& labels, int currentIndex) {
  options_ = labels;
  currentIndex_ =
      (currentIndex >= 0 && currentIndex < options_.size()) ? currentIndex : -1;

  // Batch update — mirrors Python setUpdatesEnabled(False) around clear+build
  rowsContainer_->setUpdatesEnabled(false);

  // Clear existing rows (keep trailing stretch — it's always the last item)
  while (rowsLayout_->count() > 1) {
    QLayoutItem* item = rowsLayout_->takeAt(0);
    if (QWidget* w = item ? item->widget() : nullptr) {
      w->deleteLater();
    }
    delete item;
  }
  rowWidgets_.clear();

  for (int i = 0; i < options_.size(); ++i) {
    QWidget* row = createRow(i, options_[i], i == currentIndex_, itemHeight_, itemFont_);
    rowWidgets_.push_back(row);
    // Insert before the trailing stretch (always at rowsLayout_->count()-1)
    rowsLayout_->insertWidget(rowsLayout_->count() - 1, row);
  }

  updateSize();
  rowsContainer_->setUpdatesEnabled(true);
}

QWidget* SimpleOptionsFlyout::createRow(int index, const QString& text,
                                        bool isCurrent, int itemHeight,
                                        const QFont& itemFont) {
  auto* row = new SimpleRow(index, text, isCurrent, itemHeight, itemFont,
                            rowsContainer_);
  connect(row, &SimpleRow::rowClicked, this,
          &SimpleOptionsFlyout::onRowClicked);
  return row;
}

void SimpleOptionsFlyout::updateSize(int matchWidth, bool exactMatch,
                                     std::optional<int> availableHeight) {
  const int num = options_.size();
  const int spacing = rowsLayout_->spacing();  // 2
  const QMargins outerMargins = contentLayout_->contentsMargins();
  const int marginsV = outerMargins.top() + outerMargins.bottom();

  int contentH = 0;
  int visible = 0;
  if (num == 0) {
    visible = 1;
    contentH = 50;
  } else {
    visible = std::min(num, maxVisibleItems_);
    if (availableHeight.has_value()) {
      // budget = available_height - 2 * MARGIN - margins_v
      const int budget = *availableHeight - 2 * kMargin - marginsV;
      if (budget > 0) {
        const int maxByHeight =
            std::max(1, (budget + spacing) / (itemHeight_ + spacing));
        visible = std::min(visible, maxByHeight);
      }
    }
    contentH = visible * itemHeight_ + std::max(0, visible - 1) * spacing;
  }

  const int containerH = contentH + marginsV;

  QFontMetrics fm(itemFont_);
  int maxTextWidth = 0;
  for (const auto& text : options_) {
    maxTextWidth = std::max(maxTextWidth, fm.horizontalAdvance(text));
  }

  const int finalW = maxTextWidth + 50;

  int width = 0;
  if (exactMatch && matchWidth > 0) {
    const int targetContainerWidth = std::max(1, matchWidth - kMargin * 2);
    width = std::max(finalW, targetContainerWidth);
  } else if (matchWidth > 0) {
    const int targetContainerWidth = matchWidth - kMargin * 2;
    const int minContainerWidth = std::max(180, targetContainerWidth);
    width = std::max(finalW, minContainerWidth);
  } else {
    width = std::max(finalW, 180);
  }

  // Python: self.container.setFixedSize(width, container_h)
  //         total_width = width + MARGIN * 2
  //         self.setFixedSize(total_width, container_h + MARGIN * 2)
  // In the C++ port we use Flyout (no separate container_ member visible here),
  // so we set the flyout's own fixed size to include margins.
  const int totalWidth = width + kMargin * 2;
  const int totalHeight = containerH + kMargin * 2;
  setFixedSize(totalWidth, totalHeight);
}

void SimpleOptionsFlyout::showBelow(QWidget* anchorWidget, bool exactWidthMatch) {
  if (isVisible() && anchorWidget_ == anchorWidget) {
    justOpened_ = false;
    hide();
    return;
  }

  anchorWidget_ = anchorWidget;

  if (anim_) {
    anim_->stop();
    anim_ = nullptr;
  }

  // Determine anchor width
  int anchorWidth = anchorWidget->frameGeometry().width();
  if (anchorWidth <= 0) {
    anchorWidth = anchorWidget->geometry().width();
  }
  if (anchorWidth <= 0) {
    anchorWidth = anchorWidget->width();
  }

  justOpened_ = true;
  openTimestamp_ = std::chrono::steady_clock::now();

  const int offset = kAppearExtraY - kMargin;
  const int gap = kWindowMargin;

  // Resolve host window and available rect
  QWidget* host = anchorWidget->window();
  if (parentWidget() != host) {
    setParent(host);
  }

  const QRect anchorRect(anchorWidget->mapTo(host, QPoint(0, 0)),
                         anchorWidget->size());
  const QRect avail = host->rect();

  const int spaceBelow =
      avail.bottom() - anchorRect.bottom() - offset - gap;
  const int spaceAbove =
      anchorRect.top() - avail.top() - offset - gap;
  const int budget = std::max(spaceBelow, spaceAbove);

  updateSize(anchorWidth, exactWidthMatch, budget);

  const int totalWidth = this->width();
  const int totalHeight = this->height();

  // Determine vertical position
  int finalY = 0;
  if (spaceBelow >= totalHeight || spaceBelow >= spaceAbove) {
    finalY = anchorRect.bottom() + offset;
  } else {
    finalY = anchorRect.top() - offset - totalHeight;
  }

  // Center horizontally on anchor
  const int anchorCenterX =
      anchorWidget->mapTo(host, anchorWidget->rect().center()).x();
  int finalX = static_cast<int>(anchorCenterX - totalWidth / 2.0) + 2;

  // Clamp to available rect
  finalX = std::max(avail.left(),
                    std::min(finalX, avail.right() - totalWidth));
  finalY = std::max(avail.top(),
                    std::min(finalY, avail.bottom() - totalHeight));

  const QPoint startPos(finalX, finalY - dropOffsetPx_);
  const QPoint endPos(finalX, finalY);

  move(startPos);
  setVisible(true);
  raise();
  setFocus(Qt::PopupFocusReason);

  QApplication::processEvents();

  // Enforce exact width after processEvents (mirrors Python warning + fix)
  if (exactWidthMatch && this->width() != totalWidth) {
    setFixedSize(totalWidth, totalHeight);
  }

  // Drop-in animation
  auto* animPos = new QPropertyAnimation(this, "pos", this);
  animPos->setDuration(moveDurationMs_);
  animPos->setStartValue(startPos);
  animPos->setEndValue(endPos);
  animPos->setEasingCurve(QEasingCurve::OutQuad);
  connect(animPos, &QPropertyAnimation::finished, this,
          &SimpleOptionsFlyout::onAnimationFinished);
  anim_ = animPos;
  animPos->start();
}

void SimpleOptionsFlyout::onAnimationFinished() {
  if (anim_) {
    QPropertyAnimation* a = anim_;
    anim_ = nullptr;
    a->deleteLater();
  }
}

void SimpleOptionsFlyout::onRowClicked(int idx) {
  emit itemChosen(idx);
  justOpened_ = false;
  hide();
}

void SimpleOptionsFlyout::hide() {
  Flyout::hide();
  // Python: restore focus to parent window
  if (parentWidget_) {
    if (QWidget* win = parentWidget_->window()) {
      win->activateWindow();
      win->setFocus();
    }
  }
}

void SimpleOptionsFlyout::hideEvent(QHideEvent* event) {
  // Python hideEvent: guard against closing too soon after open
  if (justOpened_) {
    using Seconds = std::chrono::duration<double>;
    const double timeSinceOpen =
        std::chrono::duration_cast<Seconds>(
            std::chrono::steady_clock::now() - openTimestamp_)
            .count();
    if (timeSinceOpen < 0.3) {
      event->ignore();
      return;
    }
    justOpened_ = false;
  }

  Flyout::hideEvent(event);

  if (anim_) {
    anim_->stop();
    anim_ = nullptr;
  }

  justOpened_ = false;

  emit closed();
}

void SimpleOptionsFlyout::ensureOverlayParent(QWidget* anchor) {
  QWidget* host = anchor ? anchor->window() : nullptr;
  if (host && parentWidget() != host) {
    setParent(host);
  }
}

}  // namespace sli::toolkit

// Required: MOC for SimpleRow defined in anonymous namespace in this .cpp
#include "simple_options_flyout.moc"
